MM-48686 : Introduce improved menus to LHS

M-ZubairAhmed opened the pull request:
#### Summary
- Adds improved menus to sidebar channel and category menu
- Made it mobile responsive, with menus in mobile opening modal

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-48686

https://mattermost.atlassian.net/browse/MM-30324
https://mattermost.atlassian.net/browse/MM-30343
https://mattermost.atlassian.net/browse/MM-33560
https://mattermost.atlassian.net/browse/MM-48263
https://mattermost.atlassian.net/browse/MM-47616
https://mattermost.atlassian.net/browse/MM-48389
https://mattermost.atlassian.net/browse/MM-48388
https://mattermost.atlassian.net/browse/MM-48879

Resolved due to the change
https://mattermost.atlassian.net/browse/MM-31345 / https://github.com/mattermost/mattermost-server/issues/16544
https://mattermost.atlassian.net/browse/MM-32400 / https://github.com/mattermost/mattermost-server/issues/16811
https://mattermost.atlassian.net/browse/MM-30084
https://mattermost.atlassian.net/browse/MM-22660
https://mattermost.atlassian.net/browse/MM-41947
https://mattermost.atlassian.net/browse/MM-45772
https://mattermost.atlassian.net/browse/MM-49963

#### Related Pull Requests
n/a

#### Screenshots
![image](https://user-images.githubusercontent.com/17708702/206169512-d816dd5d-42e9-49d8-aafb-aa64eb80d589.png)


#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Improved menus added to sidebar channel and category menus
```

M-ZubairAhmed commented:
/update-branch

mattermod commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

M-ZubairAhmed commented:
/e2e-test

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/272290) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/272290)

matthewbirtch commented:
@M-ZubairAhmed I'm going to assign UX review to @michaelgamble here. Feel free to ping me if either of you have questions though.

hmhealey commented:
Submenus also work way better now with this! MUI gives us the nice behaviour where you can move your mouse diagonally onto them without them closinng

https://user-images.githubusercontent.com/3277310/206525228-37b29360-d00c-4bef-adf1-88b085cf43a1.mov

hmhealey commented on components/menu/index.ts:
```diff
@@ -0,0 +1,8 @@
+
+export {Menu} from './menu';
+export {SubMenu} from './sub_menu';
+export {MenuItem, styled} from './menu_item';
```
Why do we need to re-export `styled` from here? If it's because of `PopoverMenuItem`, then perhaps it should live in this folder as well

M-ZubairAhmed replied:
Yes following your suggestion i moved `PopoverMenuItem` back to menu directory.

hmhealey replied:
Sounds good.

This also made me realize that, with the common components using MUI, if we want to give people using that library the ability to customize stuff, it'll be really hard to do without leaking details about MUI. Even if we didn't have to expose `styled`, we'd probably need to have them overwrite MUI's CSS classes. I'll bring it up with Los Tigres as well

hmhealey commented on components/menu/index.ts:
```diff
@@ -0,0 +1,8 @@
+
+export {Menu} from './menu';
+export {SubMenu} from './sub_menu';
+export {MenuItem, styled} from './menu_item';
+export type {MenuItemProps} from './menu_item';
+export {MenuItemDivider} from './menu_item_divider';
```
Even though this might cause some naming conflicts down the line, I do kind of like the idea of just exporting these as `Item`, `ItemProps`, and `Divider` to make them a bit more concise. That does also let people import them like this which could solve any naming conflicts
```js
import * as Menu from 'components/menu';

function MyMenu() {
    return (
        <Menu.Menu>
            <Menu.Item/>
            <Menu.Item/>
            <Menu.Divider/>
            <Menu.Item/>
            <Menu.Item/>
        </Menu.Menu>
    );
}
```

Come to think of it, that's more or less what React Bootstrap does except it does it in a way that the outer level is just `Menu` while the others are still prefixed like `Menu.Item`

M-ZubairAhmed replied:
Yes changed because i think its more concise that way.

hmhealey commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,203 @@
+    const isMenuOpen = Boolean(anchorElement);
+
+    if (isMobileView) {
+        function MenuModalComponent() {
```
I think we might want to separate this out into its own function for organization's sake. It can still live in this file though

M-ZubairAhmed replied:
separated out to its own component

hmhealey commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,203 @@
+                dispatch(closeModal(props.menuId));
+            }
+
+            function handleModalClickCapture(event: MouseEvent<HTMLUListElement>) {
```
I'm not sure if MUI's `ClickAwayListener` could be useful here or if this is a different case that it wouldn't handle

https://mui.com/material-ui/react-click-away-listener/

M-ZubairAhmed replied:
Tried the clickawaylisterner, found it not be working as intented.

hmhealey commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,203 @@
+        setAnchorElement(event.currentTarget);
+    }
+
+    function rendertriggerElement() {
```
Nit: The capitalization of this is incorrect

M-ZubairAhmed replied:
corrected

hmhealey commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,203 @@
+                aria-label={props.menuAriaLabel}
+                elevation={24}
+            >
+                <MuiMenuList
```
I don't think we need to pass our own `MenuList` into the `Menu` when using MUI. That's probably why you're having to get rid of the extra padding below this

M-ZubairAhmed replied:
Yes actually i read up a bit more, it seems MenuList would only be needed to if we are using custom Menu, so following that removed the component.

hmhealey commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,203 @@
+    '& .MuiPaper-root': {
+        backgroundColor: 'var(--center-channel-bg)',
+        boxShadow: 'var(--elevation-4) , 0 0 0 1px rgba(var(--center-channel-color-rgb), 0.08) inset',
+        minWidth: '114px',
+        maxWidth: '496px',
+        maxHeight: '80vh',
```
Are these min/max dimensions still needed? I would've figured MUI could do this for us

M-ZubairAhmed replied:
yes, these are Design/UX restrictions as per spec

hmhealey commented on components/menu/sub_menu.tsx:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    if (isMobileView) {
+        function MenuModalComponent() {
```
Same comment as with the `Menu` here. I think we could move all the mobile stuff into its own function

M-ZubairAhmed replied:
Oops missed this one. Yes moved to separate component.

hmhealey commented on components/menu/sub_menu.tsx:
```diff
@@ -0,0 +1,182 @@
+                sx={{pointerEvents: 'none'}} // disables the menu background wrapper
+                {...getOriginOfAnchorAndTransform(props.openAt)}
+            >
+                <MuiMenuList
```
And same comment here about the extra `MenuList` inside of the `Menu`. Since you're not setting `padding: 0` on this one, you're also ending up with slightly more vertical padding here
![Screen Shot 2022-12-08 at 12 45 13 PM](https://user-images.githubusercontent.com/3277310/206532271-404c6543-4355-416b-bde2-bef0669e7ee5.png)

M-ZubairAhmed replied:
Removed the extra MenuList

hmhealey commented on components/popover_menu_item/index.tsx:
```diff
@@ -0,0 +1,116 @@
+        >
+            {leadingElement}
+            <div className='label-elements'>
+                {primaryLabel}
```
Nitpicking a bit, but are these various prop names coming from the Compass design stuff? It seems a bit weird to me that we'd have separate props for `primaryLabel`/`secondaryLabel` and `subMenuDetails`/`trailingElement` when we could just have `label` and `trailingElement` props that you could pass a fragment into
```js
<PopoverMenuItem
    leadingElement={<LeadingElement/>}
    primaryLabel={<PrimaryLabel/>}
    secondaryLabel={<SecondaryLabel/>}
    subMenuDetails={<SubMenuDetails/>}
    trailingElement={<TrailingElement/>}
/>
// vs
<PopoverMenuItem
    leadingElement={<LeadingElement/>}
    label={
        <>
            <PrimaryLabel/>
            <SecondaryLabel/>
         </>
    }
    trailingElement={
        <>
            <SubMenuDetails/>
            <TrailingElement/>
        </>
    }
/>
```
Although that does end up being more lines vertically because of the fragments

M-ZubairAhmed replied:
yes simplified the props as per above suggestion.

hmhealey commented on components/sidebar/sidebar_channel/sidebar_channel_link/sidebar_channel_link.tsx:
```diff
@@ -253,23 +249,22 @@ export default class SidebarChannelLink extends React.PureComponent<Props, State
-                    closeHandler={this.props.closeHandler}
-                    onToggleMenu={this.handleMenuToggle}
-                />
+                <div className='MenuWrapper SidebarMenu'>
```
What's the menu wrapper needed for? Is it just because the trigger element needs to use some of the existing styles?

M-ZubairAhmed replied:
yes i dont like it either its for existing styling to work. Let me see if i can tweak the styled a bit.

M-ZubairAhmed replied:
removed MenuWrapper class

M-ZubairAhmed commented on components/channel_move_to_sub_menu/index.tsx:
```diff
@@ -119,51 +143,69 @@ const ChannelMoveToSubMenu = (props: Props) => {
+        <SubMenu
+            triggerId={`moveTo-${props.channel.id}`}
+            triggerElement={
+                <PopoverMenuItem
```
Yeah definitely looks like if need arises we can add SubMenuItem when we integrate in more places.

M-ZubairAhmed commented on components/compass_design_provider/index.tsx:
```diff
@@ -0,0 +1,29 @@
+    children?: ReactNode;
+}
+
+const CompassDesignProvider: FC<Props> = (props: Props) => {
```
TODO : discuss with Los Tigres team.

M-ZubairAhmed commented:
UI/UX review : https://www.figma.com/file/tbMvXozL5f8Uc8oWYTEzbh/branch/gcs1wludUgQXfXScCLGnQ5/Patterns---Popover-Menus?node-id=2722%3A85262&t=dVDtx9pjzJ67Y4aj-1

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/286924) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/286924)

matthewbirtch reviewed (changes requested):
@M-ZubairAhmed I went through everything and have a few remaining comments:

1. Can we speed up the animation to make it a bit more snappy? It looks to be set to 302ms, but can we update it to 150ms when opening (easeOut) and 100ms when closing (easeIn)?
2. When opening a menu, it sometimes automatically adds the focus state (not always though - I can't seem to find the pattern). Can we avoid this?
    <img width="495" alt="image" src="https://user-images.githubusercontent.com/2040554/211668391-af4119cf-2c8a-445e-969e-977caf1ef171.png">
3. Similarly, the hover state on menu items that open submenus seems to be using the focus state rather than the hover state
    <img width="376" alt="image" src="https://user-images.githubusercontent.com/2040554/211668710-bfd47eb6-4abe-49a9-8660-ed841dfb707b.png">
3. It looks to me like we still need to increase the spacing between the icon and the menu label - looks like we need an additional 2px of space.
4. As mentioned by @michaelgamble, the 3-dot icon that shows on a channel item while hovering should be within an icon button so that the hover state on that icon looks like the icon button that shows on a category.
    ![image](https://user-images.githubusercontent.com/2040554/211670880-b9d7ab79-5745-4e04-adc5-ef448cf6c2f2.png)
5. For the 'Leave Channel' icon, let's use `icon-exit-to-app` instead
    <img width="271" alt="image" src="https://user-images.githubusercontent.com/2040554/211671515-7de24f60-a951-4911-bfe0-29d73288b3de.png">
6. For menu items that do not use icons, it looks like we have too much padding on the left - I think because the padding is added to the left side of text rather than the right side of the icon.
    <img width="421" alt="Screenshot 2023-01-10 at 5 05 11 PM" src="https://user-images.githubusercontent.com/2040554/211672581-b5cdb29f-c9e7-4abc-ad14-22a10a7372fa.png">

M-ZubairAhmed commented:
1. Changed the enter and exit duration to mentioned one.
4. Increased the gap between leading and label element to 10px
6. Changed the icon to ExitToAppIcon
7. changed the margin element to after leading element

hmhealey commented:
@M-ZubairAhmed I found another CSS issue. The icon colour for destructive icons doesn't invert when that row is selected with the keyboard. It's correct when hovering with the mouse though.

![Screen Shot 2023-01-11 at 3 49 14 PM](https://user-images.githubusercontent.com/3277310/211915048-d7778edc-77b7-4446-989f-dff3c68d9fce.png)

---

@AshishDhama What were you doing when that happened? The only case that I think we knew that would cause it was if you opened the menu in mobile view and then resize the window to be larger without closing the menu. As far as I know, it shouldn't happen other than that.

matthewbirtch commented:
@M-ZubairAhmed I'm not seeing the changes - not sure if your commits got pushed because the test server doesn't look to be updated

M-ZubairAhmed commented:
> @M-ZubairAhmed I found another CSS issue. The icon colour for destructive icons doesn't invert when that row is selected with the keyboard. It's correct when hovering with the mouse though.
> 
> ![Screen Shot 2023-01-11 at 3 49 14 PM](https://user-images.githubusercontent.com/3277310/211915048-d7778edc-77b7-4446-989f-dff3c68d9fce.png)
> 
> @AshishDhama What were you doing when that happened? The only case that I think we knew that would cause it was if you opened the menu in mobile view and then resize the window to be larger without closing the menu. As far as I know, it shouldn't happen other than that.

fixed this

matthewbirtch reviewed (changes requested):
@M-ZubairAhmed looks like you fixed all the issues I previously identified. A few more things I'm seeing though that I hadn't seen previously.

1. The 3-dot icon that appears on categories is only supposed to show on hover (just like with channels), but it seems to always be showing here. 
    <img width="305" alt="Screenshot 2023-01-11 at 7 41 44 PM" src="https://user-images.githubusercontent.com/2040554/211948822-b8e33e14-666a-4a40-8a15-eeaeffec204f.png">

2. There seems to be something odd going on with the tooltips. After clicking away to close menus, it seems to show the tooltip of the menu trigger button that was last focused. NOTE: It only seems to do this on category ••• buttons. And it doesn't happen when I use the ESC key to close the menu
![2023-01-11 19 45 07](https://user-images.githubusercontent.com/2040554/211949311-0e8d2f79-7c05-4158-b805-e394ed508104.gif)

3. The 3-dot icon is sitting slightly too high (by 1px) and not properly centered within this button.
    <img width="405" alt="image" src="https://user-images.githubusercontent.com/2040554/211950272-469cc8f9-1cf9-4e0f-886c-0e1c53b3d23e.png">

4. The same seems to be true for channel options 3-dot icon as well.
    <img width="446" alt="image" src="https://user-images.githubusercontent.com/2040554/211950457-57db375d-6082-4681-9925-e642df828242.png">

5. Was this effort supposed to support mobile web improvements? This is how it appears for me and doesn't seem right. Is that a separate effort?
<img width="612" alt="image" src="https://user-images.githubusercontent.com/2040554/211949629-55a510c3-097d-488f-b0db-fba59d3f977d.png">

6. For the positioning issue, I was able to reproduce what @AshishDhama is seeing if I open a menu and then resize the browser to mobile and then back to desktop screen sizes. I don't think this is a common behavior though. I didn't see this happening in any other scenario.

michelengelen commented on components/menu/menu_item.tsx:
```diff
@@ -0,0 +1,208 @@
+            maxHeight: '56px',
+
+            // aria expanded to add the active styling on parent sub menu item
+            '&.Mui-active, &[aria-expanded=\'true\']': {
```
```suggestion
            '&.Mui-active, &[aria-expanded="true"]': {
```

M-ZubairAhmed replied:
Fixed

michelengelen commented on components/menu/menu_item.tsx:
```diff
@@ -0,0 +1,208 @@
+    ({isDestructive = false, hasSecondaryLabel = false}) => ({
+        '&.MuiMenuItem-root': {
+            fontFamily: '"Open Sans", sans-serif',
+            color: !isDestructive ? 'var(--center-channel-color)' : 'var(--error-text)',
```
could you turn that condition around ... strange to test for `false` instead of `true`

```suggestion
            color: isDestructive ? 'var(--error-text)' : 'var(--center-channel-color)',
```

M-ZubairAhmed replied:
I would like to prefer, primary first and secondary later.

calebroseland replied:
IMHO it's usually better to defer to ESLint, rather than introducing `eslint-disabled`s unnecessarily. A big purpose of linting is to avoid these stylistic back and forths.

M-ZubairAhmed replied:
agreed, changed the variable name and removed the eslint disabled

michelengelen commented on components/menu/menu_item.tsx:
```diff
@@ -0,0 +1,208 @@
+            flexDirection: 'row',
+            flexWrap: 'nowrap',
+            justifyContent: 'flex-start',
+            alignItems: !hasSecondaryLabel ? 'center' : 'flex-start',
```
same here:

```suggestion
            alignItems: hasSecondaryLabel ? 'flex-start' : 'center',
```

M-ZubairAhmed replied:
ref : https://github.com/mattermost/mattermost-webapp/pull/11674#discussion_r1070661919

michelengelen commented on components/sidebar/sidebar_category/sidebar_category.tsx:
```diff
@@ -322,13 +302,7 @@ export default class SidebarCategory extends React.PureComponent<Props, State> {
-                    onToggleMenu={this.handleMenuToggle}
-                />
-            );
+            categoryMenu = (<SidebarCategoryMenu category={category}/>);
```
there is no need for the parantheses anymore when single line

```suggestion
            categoryMenu = <SidebarCategoryMenu category={category}/>;
```

M-ZubairAhmed replied:
changed

michelengelen commented on components/sidebar/sidebar_category/sidebar_category_menu/sidebar_category_menu.tsx:
```diff
@@ -75,12 +83,17 @@ const SidebarCategoryMenu = (props: Props) => {
-                icon={<TrashCanOutlineIcon size={16}/>}
-                text={formatMessage({id: 'sidebar_left.sidebar_category_menu.deleteCategory', defaultMessage: 'Delete Category'})}
+                leadingElement={<TrashCanOutlineIcon size={18}/>}
+                labels={
+                    <FormattedMessage
+                        id='sidebar_left.sidebar_category_menu.deleteCategory'
+                        defaultMessage='Delete Category'
+                    />
+                }
```
in this case we should add `()`

```suggestion
                labels={(
                    <FormattedMessage
                        id='sidebar_left.sidebar_category_menu.deleteCategory'
                        defaultMessage='Delete Category'
                    />
                )}
```

M-ZubairAhmed replied:
changed

michelengelen commented on components/sidebar/sidebar_category/sidebar_category_menu/sidebar_category_menu.tsx:
```diff
@@ -96,105 +109,146 @@ const SidebarCategoryMenu = (props: Props) => {
-                icon={<PencilOutlineIcon size={16}/>}
-                text={formatMessage({id: 'sidebar_left.sidebar_category_menu.renameCategory', defaultMessage: 'Rename Category'})}
+                leadingElement={<PencilOutlineIcon size={18}/>}
+                labels={
+                    <FormattedMessage
+                        id='sidebar_left.sidebar_category_menu.renameCategory'
+                        defaultMessage='Rename Category'
+                    />
+                }
```
same here

```suggestion
                labels={(
                    <FormattedMessage
                        id='sidebar_left.sidebar_category_menu.renameCategory'
                        defaultMessage='Rename Category'
                    />
                )}
```

M-ZubairAhmed replied:
changed

michelengelen commented on components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:
```diff
@@ -31,133 +31,191 @@ type Props = PropsFromRedux & OwnProps;
-                icon={<MarkAsUnreadIcon size={16}/>}
-                text={formatMessage({id: 'sidebar_left.sidebar_channel_menu.markAsUnread', defaultMessage: 'Mark as Unread'})}
+                leadingElement={<MarkAsUnreadIcon size={18}/>}
+                labels={
+                    <FormattedMessage
+                        id='sidebar_left.sidebar_channel_menu.markAsUnread'
+                        defaultMessage='Mark as Unread'
+                    />}
```
```suggestion
                labels={(
                    <FormattedMessage
                        id='sidebar_left.sidebar_channel_menu.markAsUnread'
                        defaultMessage='Mark as Unread'
                    />
                )}
```

M-ZubairAhmed replied:
changed

michelengelen commented on components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:
```diff
@@ -31,133 +31,191 @@ type Props = PropsFromRedux & OwnProps;
     }
 
-    let favoriteMenuItem: JSX.Element | null = null;
+    let favoriteUnfavoriteMenuItem: JSX.Element | null = null;
```
naming is a bit off here, its a toggle, soooooo

```suggestion
    let favoriteMenuItemToggle: JSX.Element | null = null;
```

M-ZubairAhmed replied:
Hey, Thank you for the feedback, but i would like to leave this as is. Hope you don't mind.

michelengelen commented on components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:
```diff
@@ -31,133 +31,191 @@ type Props = PropsFromRedux & OwnProps;
-        favoriteMenuItem = (
-            <Menu.ItemAction
+        favoriteUnfavoriteMenuItem = (
+            <Menu.Item
```
this whole condition looks like it can be changed to only alter the props.

I would much rather prefer a style like this:
```typescript
// handleUnfavoriteChannel and handleFavoriteChannel could be ideally consolidated
const makeToggleFavoriteChannelHandler = (isFavorite) => isFavorite ? handleUnfavoriteChannel : handleFavoriteChannel;
const favoriteMenuItemToggle = (
    <Menu.Item
        id={`unfavorite-${props.channel.id}`}
        onClick={makeToggleFavoriteChannelHandler(props.isFavorite)}
        leadingElement={props.isFavorite ? <StarIcon size={18}/> : <StarOutlineIcon size={18}/>}
        labels={props.isFavorite ? (
            <FormattedMessage
                id='sidebar_left.sidebar_channel_menu.unfavoriteChannel'
                defaultMessage='Unfavorite'
            />
        ) : (
            <FormattedMessage
                id='sidebar_left.sidebar_channel_menu.favoriteChannel'
                defaultMessage='Favorite'
            />
        )}
    />
);
```

M-ZubairAhmed replied:
Hey, Thank you for the feedback, but i would like to leave this as is. Hope you don't mind.

michelengelen replied:
In this case I actually do mind ... the code is way more readable and you safe a considerable amount of lines

hmhealey replied:
I agree that it's less lines, but I actually disagree about it being easier to read. I'd actually argue that taking more lines for it makes it easier to read since you can clearly see the code for each case in its own section rather than needing to parse which parts of each line occur in each case. That's why I tend to rarely use ternary statements and prefer if statements for anything that's even slightly complex.

In this case in particular,  none of the props that we're passing into the `Menu.Item` are actually shared between the favourite/unfavourite menu items even though they have a similar function for end users

AshishDhama replied:
I have a question here, does it make any difference performance-wise?

michelengelen commented on components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:
```diff
@@ -31,133 +31,191 @@ type Props = PropsFromRedux & OwnProps;
 
-        muteChannelMenuItem = (
-            <Menu.ItemAction
+        muteUnmuteChannelMenuItem = (
```
same structure as above would also clean this up nicely

M-ZubairAhmed replied:
Hey, Thank you for the feedback, but i would like to leave this as is. Hope you don't mind.

michelengelen replied:
same as above

michelengelen commented on components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:
```diff
@@ -31,133 +31,191 @@ type Props = PropsFromRedux & OwnProps;
 
 const SidebarChannelMenu = (props: Props) => {
     const isLeaving = useRef(false);
```
not part of this, but with this pattern you can safe some chars

```suggestion
    const {current: isLeaving} = useRef(false);
    ...
    // just use like this
    isLeaving = true;
```

works best with primitive values! 🚀

M-ZubairAhmed replied:
Hey, Thank you for the feedback, but i would like to leave this as is. Hope you don't mind.

michelengelen reviewed (changes requested):
nice work @M-ZubairAhmed ... a few comments. I like your inline comenting style! 👍🏼

M-ZubairAhmed commented:
@matthewbirtch 
1. Fixed it to show only on hover state
2. Fixed the tooltip bug
3. Fixed icon alignment
4. Fixed icon alignment
5. Yes the menu now open as modals in mobile

matthewbirtch commented:
Looks like everything is resolved now. Just a question about this one

> 5\. Yes the menu now open as modals in mobile

Is there design work for this? Is this a temporary solution? Was UX involved in this approach? Sorry, this is new to me. The most recent design work for mobile web was to make it more of a bottom sheet. But maybe I wasn't a part of some conversations.

M-ZubairAhmed commented:
> Looks like everything is resolved now. Just a question about this one
> 
> > 5. Yes the menu now open as modals in mobile
> 
> Is there design work for this? Is this a temporary solution? Was UX involved in this approach? Sorry, this is new to me. The most recent design work for mobile web was to make it more of a bottom sheet. But maybe I wasn't a part of some conversations.

Mobile web view is not a priority since it will be soon sun setted. Having the drawer menu for mobile web view required a lot of work for something temporarily. So we decided to go with modals for now. They are not drawer menu as per design but are improvement to whats we got right now.

<img width="574" alt="image" src="https://user-images.githubusercontent.com/17708702/212133311-0b277a3a-d999-4a05-a9c7-2bf43d4d243f.png">

matthewbirtch reviewed (approved):
@M-ZubairAhmed as discussed, a few minor updates to the mobile modals to make it a little nicer and we're good:
- add negative margin on the menu list so that there isn't so much space on the left and right side on the mobile menus
- hide 'x' button from the top-right corner of the mobile menu modals, make sure clicking away from the menus or clicking on the darkened background layer will dismiss them - this will also prevent collisions between the close button and the menu items below

I'm going to approve this now to unblock. It's looking/working great!

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/288237) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/288237)

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/288251) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/288251)

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/288577) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/288577)

jgilliam17 commented:
[E2e](https://mattermost-cypress-report.s3.amazonaws.com/288577-ff16a98-onprem-ent-webapp-pr-11674-mm-ee-test:ff16a98/mochawesome.html) report has few failures not listed on master:
Direct message redirect
Close GM
Channel routing
Interactive menu
Message reply (older message)

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/288783) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/288783)

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/288882) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/288882)

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/288890) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/288890)

M-ZubairAhmed commented:
@jgilliam17 
All the e2e related to this PR are fixed now in the latest report

Following are the summary of the failed test
- 5141 : failed on master
- 660_4 : failed on master
- 642 : flaky, sometimes fails on master
- 1229 : failed on master
- 1230 : failed on master
- 1805 : flaky, failed on master, passed locally
- 1960 : failed on master
- 1032 : flaky

M-ZubairAhmed commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,246 @@
+        dispatch(closeModal(props.menuId));
+    }
+
+    function handleModalClickCapture(event: MouseEvent<HTMLDivElement>) {
```
ref : https://github.com/mattermost/mattermost-webapp/pull/11674#discussion_r1046786673 Looks like improvement which can be done later

M-ZubairAhmed commented on components/channel_move_to_sub_menu/index.tsx:
```diff
@@ -74,38 +76,63 @@ const ChannelMoveToSubMenu = (props: Props) => {
+                );
+            }
+
+            let selectedCategory = null;
+            if (currentCategory && currentCategory.display_name === category.display_name) {
+                selectedCategory = (
+                    <CheckIcon
+                        color='var(--button-bg)'
+                        size={18}
+                    />
+                );
```
Hey, Thank you for the feedback, but i would like to leave this as is. Hope you don't mind.

M-ZubairAhmed commented on components/channel_move_to_sub_menu/index.tsx:
```diff
@@ -56,14 +54,18 @@ const ChannelMoveToSubMenu = (props: Props) => {
     });
 
-    function handleMoveToCategory(categoryId: string) {
+    function handleMoveToCategory(event: MouseEvent<HTMLLIElement>, categoryId: string) {
```
Hey, Thank you for the feedback, but i would like to leave this as is. Hope you don't mind.

M-ZubairAhmed commented on components/channel_move_to_sub_menu/index.tsx:
```diff
@@ -74,38 +76,63 @@ const ChannelMoveToSubMenu = (props: Props) => {
+                    leadingElement={category.type === CategoryTypes.FAVORITES ? (<StarOutlineIcon size={18}/>) : (<FolderOutlineIcon size={18}/>)}
+                    labels={text}
+                    trailingElements={selectedCategory}
+                    onClick={(event) => handleMoveToCategory(event, category.id)}
```
Hey, Thank you for the feedback, but i would like to leave this as is. Hope you don't mind.

M-ZubairAhmed commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,246 @@
+                    id: props.menuId,
+                    'aria-label': props.menuAriaLabel,
+                }}
+                TransitionProps={{
+                    mountOnEnter: true,
+                    unmountOnExit: true,
+                    timeout: {
+                        enter: MENU_OPEN_ANIMATION_DUREATION,
+                        exit: MENU_CLOSE_ANIMATION_DUREATION,
+                    },
+                }}
```
I am not sure of the animation functions here. but the current animation seems ok by UX, so i will leave it to the Los tigres team to make changes later.

M-ZubairAhmed commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,246 @@
+    function handleMenuKeyDown(event: KeyboardEvent<HTMLDivElement>) {
+        if (event.key === 'Enter') {
+            const target = event.target as HTMLElement;
+            const ariaHasPopup = target?.getAttribute('aria-haspopup') === 'true' ?? false;
```
Lets leave it as is, i always prefer to fallback when using optional chaining to specific it to boolean value here instead of boolean or undefined.

M-ZubairAhmed replied:
changed

M-ZubairAhmed commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,246 @@
+import {MuiMenuStyled} from './menu_styled';
+
+const OVERLAY_TIME_DELAY = 500;
+const MENU_OPEN_ANIMATION_DUREATION = 150;
+const MENU_CLOSE_ANIMATION_DUREATION = 100;
```
fixed

M-ZubairAhmed commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,246 @@
+    menuButtonTooltipId?: string;
+    menuButtonTooltipPlacement?: 'top' | 'bottom' | 'left' | 'right';
+    menuButtonTooltipText?: string;
+    menuButtonTooltipClassName?: string;
```
yes this looks much cleaner, changed

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/289344) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/289344)

M-ZubairAhmed commented:
> @M-ZubairAhmed as discussed, a few minor updates to the mobile modals to make it a little nicer and we're good:
> 
> * add negative margin on the menu list so that there isn't so much space on the left and right side on the mobile menus
> * hide 'x' button from the top-right corner of the mobile menu modals, make sure clicking away from the menus or clicking on the darkened background layer will dismiss them - this will also prevent collisions between the close button and the menu items below
> 
> I'm going to approve this now to unblock. It's looking/working great!

@matthewbirtch 
<img width="366" alt="image" src="https://user-images.githubusercontent.com/17708702/212636545-1f17fdf5-aed7-49fb-b8e2-85b2b1bcfb26.png">

AshishDhama commented on components/channel_header_dropdown/channel_header_dropdown_items.tsx:
```diff
@@ -11,7 +11,7 @@ import {isGuest} from 'mattermost-redux/utils/user_utils';
 import MobileChannelHeaderPlug from 'plugins/mobile_channel_header_plug';
 
-import ChannelMoveToSubMenu from 'components/channel_move_to_sub_menu';
+import ChannelMoveToSubMenuOld from 'components/channel_move_to_sub_menu_old';
```
If I am not wrong, I think it's industry practice to rename things with the Legacy prefix or suffix in a scenario like this so maybe `ChannelMoveToSubMenuLegacy`. No need to change anything here.

AshishDhama reviewed (approved):
Great Job, @M-ZubairAhmed

AshishDhama commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,259 @@
+                open={isMenuOpen}
+                onClose={handleMenuClose}
+                onClick={handleMenuClick}
+                onKeyDown={handleMenuKeyDown}
```
Can we make sure onKeyDown can be called from outside as well? In the dot menu component, I am thinking of using it for keyboard shortcuts instead of adding an event listener
<img width="568" alt="Screenshot 2023-01-17 at 12 13 50 AM" src="https://user-images.githubusercontent.com/16203333/212747282-47e2a05c-d6e6-4943-8227-9bc6f043d7d4.png">

AshishDhama replied:
When we add an event listener, events are from dom. But this they remain React synthetic events

M-ZubairAhmed replied:
This is interesting, let me verify that before we merge the pr, but we can always improve on this component as we integrate it in more places.

AshishDhama commented:
@M-ZubairAhmed We don't have anything for this new label
<img width="290" alt="Screenshot 2023-01-17 at 12 58 52 AM" src="https://user-images.githubusercontent.com/16203333/212753364-8684f67e-d469-4395-ba07-30b5d76ebe8a.png">

AshishDhama commented on components/menu/menu_item.tsx:
```diff
@@ -0,0 +1,208 @@
+    } = props;
+
+    // When both primary and secondary labels are passed, we need to apply minor changes to the styling. Check below in styled component for more details.
+    const hasSecondaryLabel = labels && labels.props && labels.props.children && Children.count(labels.props.children) === 2;
```
I am already facing a scenario where this is creating an issue. It would be better to let the user decide.
<img width="290" alt="Screenshot 2023-01-17 at 12 58 52 AM" src="https://user-images.githubusercontent.com/16203333/212754693-1938e44c-b59e-44b0-8aef-62fb69d4097d.png">

M-ZubairAhmed replied:
It wasn't there in the design docs, so it got missed

michelengelen replied:
I have to agree with @AshishDhama here. We should let the user decide what to put into that label section.

michelengelen commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,259 @@
+                    overlay={
+                        <Tooltip
+                            id={props.menuButtonTooltip.id}
+                            className={props.menuButtonTooltip?.class ?? ''}
```
`??` not needed here

M-ZubairAhmed replied:
i didnt want it to have undefined in className hence ??

calebroseland replied:
Whether an empty string or undefined, it has the same outcome.

michelengelen commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,259 @@
+                disableAutoFocusItem={disableAutoFocusItem} // This is not anti-pattern, see handleMenuButtonMouseDown
+                MenuListProps={{
+                    id: props.menu.id,
+                    'aria-label': props.menu?.['aria-label'] ?? '',
```
`??` not needed here

M-ZubairAhmed replied:
i am not how the MenuListProps handles aria-label undefined. for normal components i would have just done that, but for this i had to do this

michelengelen commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    useEffect(() => {
+        if (props.menu.onToggle) {
+            props.menu.onToggle(isMenuOpen);
+        }
```
```suggestion
        props.menu?.onToggle?.(isMenuOpen);
```

michelengelen commented on components/menu/menu_item.tsx:
```diff
@@ -0,0 +1,208 @@
+            },
+
+            '&>.leading-element': {
+                width: '18px',
+                height: '18px',
```
since this will be an `Icon` this should be handled by the IconComponent

michelengelen commented on components/menu/sub_menu.tsx:
```diff
@@ -0,0 +1,240 @@
+    };
+
+    if (isMobileView) {
+        return (<ParentMenuItem {...passedInTriggerButtonProps}/>);
```
```suggestion
        return <ParentMenuItem {...passedInTriggerButtonProps}/>;
```

M-ZubairAhmed replied:
throws eslint error without the bracket

michelengelen replied:
not for me

michelengelen replied:
but its not-blocking, so I'll let that slip

michelengelen commented on components/sidebar/sidebar_category/sidebar_category_menu/sidebar_category_menu.tsx:
```diff
@@ -96,105 +112,167 @@ const SidebarCategoryMenu = (props: Props) => {
+        />
+    );
+
+    function handleMenuToggle(isOpen: boolean) {
```
nit: `toggleMenuOpen`

michelengelen commented on components/sidebar/sidebar_category/sidebar_category_menu/sidebar_category_menu.tsx:
```diff
@@ -96,105 +112,167 @@ const SidebarCategoryMenu = (props: Props) => {
+            className={classNames(
+                'SidebarMenu',
+                'MenuWrapper',
+                {menuOpen: isMenuOpen},
+                {'MenuWrapper--open': isMenuOpen},
```
if the condition is the same you can just use one extra class on it
```suggestion
                {menuOpen: isMenuOpen},
```

M-ZubairAhmed replied:
changed

michelengelen commented on components/sidebar/sidebar_category/sidebar_category_sorting_menu/sidebar_category_sorting_menu.tsx:
```diff
@@ -55,99 +113,108 @@ const SidebarCategorySortingMenu = (props: Props) => {
+            id={`openDirectMessage-${props.category.id}`}
+            onClick={props.handleOpenDirectMessagesModal}
+            leadingElement={<AccountPlusOutlineIcon size={18}/>}
+            labels={
+                <FormattedMessage
+                    id='sidebar.openDirectMessage'
+                    defaultMessage='Open a direct message'
+                />
+            }
```
```suggestion
            labels={(
                <FormattedMessage
                    id='sidebar.openDirectMessage'
                    defaultMessage='Open a direct message'
                />
            )}
```

M-ZubairAhmed replied:
changed

michelengelen commented on components/sidebar/sidebar_channel/sidebar_channel_link/sidebar_channel_link.tsx:
```diff
@@ -250,14 +252,22 @@ export default class SidebarChannelLink extends React.PureComponent<Props, State
+                    className={classNames(
+                        'SidebarMenu',
+                        'MenuWrapper',
+                        {menuOpen: this.state.isMenuOpen},
+                        {'MenuWrapper--open': this.state.isMenuOpen},
```
```suggestion
                        {menuOpen: this.state.isMenuOpen},
```

jgilliam17 commented:
Thanks @M-ZubairAhmed 
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/289344-0f9e34b-onprem-ent-webapp-pr-11674-mm-ee-test:0f9e34b/mochawesome.html) looks good 
Apologies for starting from mobile webview, I ended up testing first from your latest comment.

Seeing few issues:
1. Mobile webview: in Safari and Chrome - Unable to dismiss menu by clicking below, only dismisses on above modal outside click
2. Mobile webview: 3-dot menus are often unresponsive on click

https://user-images.githubusercontent.com/52937121/212977603-048efdfa-4355-4d12-afd4-897044654894.mov

3. Mobile webview: I expected modal to be dismissed on Mute, Favorite and Mark as unread. On test server menu stays open.
4. Mobile webview: Empty team sidebar when user belongs to only one team
5. Focus/cursor missing in Create New category when opening from channel menu>move to>new category. Also in mobile webview.
6. Focus/cursor missing in Add Member to channel when opening from channel menu>Add Members> Add people. Also in mobile webview.
7. Console error ^ probably related to step 5.  above. Reproduces when opening create new category
<img width="1637" alt="Screen Shot 2023-01-17 at 1 15 03 PM" src="https://user-images.githubusercontent.com/52937121/212979166-25bec2f7-7bf8-42d9-9826-04287c179e56.png">

matthewbirtch commented:
@M-ZubairAhmed one minor thing I missed - it looks like the 1px border/box-shadow on the menu container needs to be updated to `0.16` opacity of the center-channel-text color (rather than `0.08`). Sorry for missing that!

M-ZubairAhmed commented:
/update-branch

AshishDhama commented on components/menu/menu_item.tsx:
```diff
@@ -0,0 +1,208 @@
+            },
+
+            '&.Mui-focusVisible': {
+                boxShadow: !isDestructive ? '0 0 0 2px var(--denim-sidebar-active-border) inset' : '0 0 0 2px rgba(var(--button-color-rgb), 0.16) inset',
```
@M-ZubairAhmed, I missed it earlier, this should be 
```suggestion
                boxShadow: !isDestructive ? '0 0 0 2px var(--sidebar-text-active-border) inset' : '0 0 0 2px rgba(var(--button-color-rgb), 0.16) inset',
```

M-ZubairAhmed replied:
changed

M-ZubairAhmed commented:
Hi @jgilliam17 
1. This is an existing issue with any generic modals - https://mattermost.atlassian.net/browse/MM-49937
2. This is now fixed
3. Modal now closes on menu items
4. This is an existing issue, https://mattermost.atlassian.net/browse/MM-49938
5. Let me create follow up tickets for this
6. ^
7. This is now fixed

jgilliam17 commented:
/update-branch

jgilliam17 commented:
Thanks @M-ZubairAhmed for adding follow up tickets.
re-tested, LGTM - pending e2es.

jgilliam17 commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/293038) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/293038)

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/293183) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/293183)

M-ZubairAhmed commented on components/menu/sub_menu.tsx:
```diff
@@ -0,0 +1,240 @@
+    forceOpenOnLeft?: boolean; // Most of the times this is not needed, since submenu position is calculated and placed
+
+    children: ReactNode;
+
```
fixed

M-ZubairAhmed commented:
@jgilliam17 the focus bugs are fixed too

M-ZubairAhmed commented on components/sidebar/sidebar_category/sidebar_category_sorting_menu/sidebar_category_sorting_menu.tsx:
```diff
@@ -55,99 +113,108 @@ const SidebarCategorySortingMenu = (props: Props) => {
+                <Menu.Item
+                    id={`showDmCount-${props.category.id}-${dmGmShowCount}`}
+                    key={`showDmCount-${props.category.id}-${dmGmShowCount}`}
+                    labels={<span>{dmGmShowCount}</span>}
```
without span it throws type error

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/293269) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/293269)

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
@jgilliam17 QA report looks good.
All failures in the report are also part of latest master e2e report except for
`MM-T924 Users - Page through users list` which i think could just be flaky

calebroseland commented on components/menu/menu.tsx:
```diff
@@ -0,0 +1,264 @@
+                        menuId: props.menu.id,
+                        menuAriaLabel: props.menu?.['aria-label'] ?? '',
+                        onModalClose: handleMenuModalClose,
+                        children: props.children,
```
@M-ZubairAhmed I'm seeing serialization warnings and crashes when dispatching this and redux devtools are attached.

<img width="965" alt="CleanShot 2023-01-25 at 11 22 55@2x" src="https://user-images.githubusercontent.com/11724372/214656703-5bf54b55-ee77-478b-b3dc-ad67f33ce6b9.png">
<img width="204" alt="CleanShot 2023-01-25 at 11 23 05@2x" src="https://user-images.githubusercontent.com/11724372/214656704-fcbd03fb-97ed-484f-9c50-452bd18f5997.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/11724372/214669457-fccc5444-737f-494a-b174-c4d62b0ac024.gif">


Is there a way to avoid this? I know it'd be ideal to use `openModal`/ModalController, but if mounting it manually would avoid the mem spike in redux state, maybe it would be worth it to avoid.

M-ZubairAhmed replied:
As discussed this is an edge case where this will happen when redux dev tools are open in mobile web view. Opened a ticket https://mattermost.atlassian.net/browse/MM-50006

jgilliam17 commented:
Thanks @M-ZubairAhmed 
I verified last 2 focus issues are also fixed. LGTM

michelengelen commented on components/menu/sub_menu.tsx:
```diff
@@ -0,0 +1,239 @@
+                    aria-label={menuAriaLabel}
+                    className={A11yClassNames.POPUP}
+                    onKeyDown={handleSubMenuKeyDown}
+                    sx={{
+                        pointerEvents: 'auto', // reset pointer events to default from here on
+                        paddingTop: 0,
+                        paddingBottom: 0,
+                    }}
```
nit: its better to define those values beforehand to reduce unnecessary rerenders

michelengelen reviewed (approved):
I'll approve since it works as expected ... the code-style and readability can be improved. Would you mind creating a follow-up ticket for that?

This is to be expected from a PR this big! 🚀

calebroseland reviewed (approved):
🚀
